database2/ffboltdb/blockio.go

```diff
@@ -0,0 +1,745 @@
+// will also lead to any data which failed to be undone being overwritten and
+// thus behaves as desired.
+//
+// Therefore, any errors are simply logged at a warning level rather than being
+// returned since there is nothing more that could be done about it anyways.
+func (s *blockStore) handleRollback(oldBlockFileNum, oldBlockOffset uint32) {
+	// Grab the write cursor mutex since it is modified throughout this
+	// function.
+	wc := s.writeCursor
+	wc.Lock()
+	defer wc.Unlock()
+
+	// Nothing to do if the rollback point is the same as the current write
+	// cursor.
+	if wc.curFileNum == oldBlockFileNum && wc.curOffset == oldBlockOffset {
+		return
+	}
+
+	// Regardless of any failures that happen below, reposition the write
+	// cursor to the old block file and offset.
+	defer func() {
+		wc.curFileNum = oldBlockFileNum
+		wc.curOffset = oldBlockOffset
+	}()
+
+	log.Debugf("ROLLBACK: Rolling back to file %d, offset %d",
+		oldBlockFileNum, oldBlockOffset)
+
+	// Close the current write file if it needs to be deleted.  Then delete
+	// all files that are newer than the provided rollback file while
+	// also moving the write cursor file backwards accordingly.
+	if wc.curFileNum > oldBlockFileNum {
+		wc.curFile.Lock()
+		if wc.curFile.file != nil {
+			_ = wc.curFile.file.Close()
+			wc.curFile.file = nil
+		}
+		wc.curFile.Unlock()
+	}
+	for ; wc.curFileNum > oldBlockFileNum; wc.curFileNum-- {
+		if err := s.deleteFileFunc(wc.curFileNum); err != nil {
+			_ = log.Warnf("ROLLBACK: Failed to delete block file "+
+				"number %d: %v", wc.curFileNum, err)
+			return
+		}
+	}
+
+	// Open the file for the current write cursor if needed.
+	wc.curFile.Lock()
+	if wc.curFile.file == nil {
+		obf, err := s.openWriteFileFunc(wc.curFileNum)
+		if err != nil {
+			wc.curFile.Unlock()
+			_ = log.Warnf("ROLLBACK: %v", err)
+			return
+		}
+		wc.curFile.file = obf
+	}
+
+	// Truncate the to the provided rollback offset.
+	if err := wc.curFile.file.Truncate(int64(oldBlockOffset)); err != nil {
+		wc.curFile.Unlock()
+		_ = log.Warnf("ROLLBACK: Failed to truncate file %d: %v",
+			wc.curFileNum, err)
+		return
+	}
+
+	// Sync the file to disk.
+	if err := wc.curFile.file.Sync(); err != nil {
+		wc.curFile.Unlock()
+		_ = log.Warnf("ROLLBACK: Failed to sync file %d: %v",
+			wc.curFileNum, err)
+		return
+	}
+	wc.curFile.Unlock()
+	return
+}
+
+// scanBlockFiles searches the database directory for all flat block files to
+// find the end of the most recent file.  This position is considered the
+// current write cursor which is also stored in the metadata.  Thus, it is used
+// to detect unexpected shutdowns in the middle of writes so the block files
+// can be reconciled.
+func scanBlockFiles(dbPath string) (int, uint32) {
+	lastFile := -1
+	fileLen := uint32(0)
+	for i := 0; ; i++ {
+		filePath := blockFilePath(dbPath, uint32(i))
+		st, err := os.Stat(filePath)
+		if err != nil {
+			break
```

Comment:
might want to check for `os.IsNotExist` here.  Other errors might need to be returned.